osquery/tables/system/tests/posix/ssh_keys_tests.cpp

```diff
@@ -0,0 +1,283 @@
+        std::ofstream(filepath.native(), std::ios::out | std::ios::binary);
+    fout << "this is an invalid key" << '\n';
+  }
+  auto const uid = std::to_string(geteuid());
+  genSSHkeyForHosts(
+      uid, std::to_string(getegid()), directory.native(), results);
+  ASSERT_EQ(results.size(), 0u);
+}
+
+TEST_F(SshKeysTests, rsa_key_unencrypted) {
+  auto results = QueryData{};
+
+  auto filepath = ssh_directory / fs::path("rsa_unencrypted");
+  {
+    auto fout =
+        std::ofstream(filepath.native(), std::ios::out | std::ios::binary);
+    fout << kRsaUnencrypted;
+  }
+  auto const uid = std::to_string(geteuid());
+  genSSHkeyForHosts(
+      uid, std::to_string(getegid()), directory.native(), results);
+  ASSERT_EQ(results.size(), 1u);
+
+  const auto& row = results[0];
+  EXPECT_EQ(row.at("uid"), uid);
+  EXPECT_EQ(row.at("path"), fs::canonical(filepath).native());
+  EXPECT_EQ(row.at("encrypted"), "0");
+  EXPECT_EQ(row.at("key_type"), "rsa");
+}
+
+TEST_F(SshKeysTests, rsa_key_encrypted) {
+  auto results = QueryData{};
+
+  auto filepath = ssh_directory / fs::path("rsa_unencrypted");
+  {
+    auto fout =
+        std::ofstream(filepath.native(), std::ios::out | std::ios::binary);
+    fout << kRsaEncrypted;
+  }
+  auto const uid = std::to_string(geteuid());
+  genSSHkeyForHosts(
+      uid, std::to_string(getegid()), directory.native(), results);
+  ASSERT_EQ(results.size(), 1u);
+
+  const auto& row = results[0];
+  EXPECT_EQ(row.at("uid"), uid);
+  EXPECT_EQ(row.at("path"), fs::canonical(filepath).native());
+  EXPECT_EQ(row.at("encrypted"), "1");
+  EXPECT_EQ(row.at("key_type"), "");
+}
+
+TEST_F(SshKeysTests, dsa_unencrypted) {
+  auto results = QueryData{};
+
+  auto filepath = ssh_directory / fs::path("dsa_unencrypted");
+  {
+    auto fout =
+        std::ofstream(filepath.native(), std::ios::out | std::ios::binary);
+    fout << kDsaUnencrypted;
+  }
+  auto const uid = std::to_string(geteuid());
+  genSSHkeyForHosts(
+      uid, std::to_string(getegid()), directory.native(), results);
+  ASSERT_EQ(results.size(), 1u);
+
+  const auto& row = results[0];
+  EXPECT_EQ(row.at("uid"), uid);
+  EXPECT_EQ(row.at("path"), fs::canonical(filepath).native());
+  EXPECT_EQ(row.at("encrypted"), "0");
+  EXPECT_EQ(row.at("key_type"), "dsa");
+}
+
+TEST_F(SshKeysTests, dsa_encrypted) {
+  auto results = QueryData{};
+
+  auto filepath = ssh_directory / fs::path("dsa_encrypted");
+  {
+    auto fout =
+        std::ofstream(filepath.native(), std::ios::out | std::ios::binary);
+    fout << kDsaEncrypted;
+  }
+  auto const uid = std::to_string(geteuid());
+  genSSHkeyForHosts(
+      uid, std::to_string(getegid()), directory.native(), results);
+  ASSERT_EQ(results.size(), 1u);
+
+  const auto& row = results[0];
+  EXPECT_EQ(row.at("uid"), uid);
+  EXPECT_EQ(row.at("path"), fs::canonical(filepath).native());
+  EXPECT_EQ(row.at("encrypted"), "1");
+  EXPECT_EQ(row.at("key_type"), "");
+}
+
+TEST_F(SshKeysTests, ed25519_unencrypted) {
```

Comment:
These ed25519 tests are currently failing because the version of openssl that we bundle doesn't parse them properly. We can either update the openssl version or fall back to the brittle ssh key parsing code from the previous revision.

Reply:
Could we keep the current implementation in the PR but check https://www.openssl.org/docs/manmaster/man3/OPENSSL_VERSION_NUMBER.html within the tests and only run this test for the correct version?

Reply:
Turns out the new version doesn't work either! Now we use openssl if possible, but if it can't read the key we fall back to naive parsing.